server/lib/chat/chat-service.ts

```diff
@@ -340,45 +341,46 @@ export default class ChatService {
       beforeTime && beforeTime > -1 ? new Date(beforeTime) : undefined,
     )
 
-    const messages = dbMessages.map(m => {
-      switch (m.data.type) {
+    const messages: ServerChatMessage[] = []
+    const users: SbUser[] = []
+    let mentionIds = new global.Set<SbUserId>()
+
+    for (const msg of dbMessages) {
+      switch (msg.data.type) {
         case ServerChatMessageType.TextMessage:
-          return {
-            id: m.msgId,
-            type: m.data.type,
-            channel: m.channelName,
-            from: m.userId,
-            user: {
-              id: m.userId,
-              name: m.userName,
-            },
-            time: Number(m.sent),
-            text: m.data.text,
-          }
+          messages.push({
+            id: msg.msgId,
+            type: msg.data.type,
+            channel: msg.channelName,
+            from: msg.userId,
+            time: Number(msg.sent),
+            text: msg.data.text,
+          })
+          users.push({ id: msg.userId, name: msg.userName })
+          mentionIds = new global.Set<SbUserId>([...mentionIds, ...(msg.data.mentions || [])])
```

Comment:
so when I was saying to use the built-in Set I didn't mean to also use it as if it were immutable :P

This is gonna be horribly inefficient, just mutate the existing set